File: utilities/UI.py
import os
import time
from typing import Any


class Colors:
    """ANSI color codes for terminal output"""
    RED = '\033[91m'
    GREEN = '\033[92m'
    YELLOW = '\033[93m'
    BLUE = '\033[94m'
    MAGENTA = '\033[95m'
    CYAN = '\033[96m'
    WHITE = '\033[97m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'
    END = '\033[0m'
    GOLD = '\033[93m'
    ORANGE = '\033[38;5;208m'
    PURPLE = '\033[95m'
    DARK_GRAY = '\033[90m'
    LIGHT_GRAY = '\033[37m'
    GRAY = '\033[90m'

    # Rarity colors for items
    COMMON = '\033[37m'
    UNCOMMON = '\033[92m'
    RARE = '\033[94m'
    EPIC = '\033[95m'
    LEGENDARY = '\033[93m'

    @staticmethod
    def _color(code: str) -> str:
        return code

    @classmethod
    def wrap(cls, text: str, color_code: str) -> str:
        return f"{color_code}{text}{cls.END}"


def clear_screen():
    """Clear the terminal screen."""
    time.sleep(1)
    command = 'cls' if os.name == 'nt' else 'clear'
    os.system(command)


def create_progress_bar(current: int,
                        maximum: int,
                        width: int = 20,
                        color: str = Colors.GREEN) -> str:
    """Create a visual progress bar."""
    if maximum <= 0:
        return "[" + " " * width + "]"
    filled_width = int((current / maximum) * width)
    filled = "█" * filled_width
    empty = "░" * (width - filled_width)
    percentage = (current / maximum) * 100
    return f"[{Colors.wrap(filled, color)}{empty}] {percentage:.1f}%"


def create_separator(char: str = "=", length: int = 60) -> str:
    """Create a visual separator line."""
    return char * length


def create_section_header(title: str, char: str = "=", width: int = 60) -> str:
    """Create a decorative section header."""
    padding = (width - len(title) - 2) // 2
    header_text = f"{char * padding} {title} {char * padding}"
    return Colors.wrap(header_text, f"{Colors.CYAN}{Colors.BOLD}")


def display_welcome_screen(lang: Any, game_instance: Any):
    """Display welcome screen and handle menu."""
    from main import ask
    while True:
        clear_screen()
        print(f"{Colors.CYAN}{Colors.BOLD}")
        print("=" * 60)
        print(f"             {lang.get('game_title_display')}")
        print(f"       {lang.get('game_subtitle_display')}")
        print("=" * 60)
        print(f"{Colors.END}")
        print(lang.get("welcome_message"))
        print(
            "Choose your path wisely, for every decision shapes your destiny.\n"
        )
        print(
            f"{Colors.BOLD}{Colors.CYAN}=== {lang.get('main_menu')} ==={Colors.END}"
        )
        print(f"{Colors.CYAN}1.{Colors.END} {lang.get('new_game')}")
        print(f"{Colors.CYAN}2.{Colors.END} {lang.get('load_game')}")
        print(f"{Colors.CYAN}3.{Colors.END} {lang.get('settings')}")
        print(f"{Colors.CYAN}4.{Colors.END} {lang.get('mods')}")
        print(f"{Colors.CYAN}5.{Colors.END} {lang.get('quit')}\n")

        choice = ask(f"{Colors.CYAN}Choose an option (1-5): {Colors.END}")
        if choice == "1":
            return "new_game"
        if choice == "2":
            return "load_game"
        if choice == "3":
            game_instance.settings_welcome()
        if choice == "4":
            game_instance.mods_welcome()
        if choice == "5":
            print(lang.get("thank_exit"))
            clear_screen()
            import sys
            sys.exit(0)


def display_main_menu(lang: Any, player: Any, area_name: str, menu_max: str):
    """Display the main game menu options."""
    from utilities.battle import create_hp_mp_bar
    print(f"\n{Colors.BOLD}=== {lang.get('main_menu')} ==={Colors.END}")
    print(lang.get("current_location", area=area_name))

    # Time and weather
    display_hour = int(player.hour)
    display_minute = int((player.hour - display_hour) * 60)
    time_str = lang.get("current_time",
                        hour=f"{display_hour:02d}:{display_minute:02d}")
    day_str = lang.get("current_day", day=str(player.day))
    weather_desc = player.get_weather_description(lang)
    print(f"{Colors.YELLOW}{time_str} | {day_str}{Colors.END}")
    print(f"{Colors.CYAN}{weather_desc}{Colors.END}")

    print(f"{Colors.CYAN}1.{Colors.END} {lang.get('explore')}")
    print(f"{Colors.CYAN}2.{Colors.END} {lang.get('view_character')}")
    print(f"{Colors.CYAN}3.{Colors.END} {lang.get('travel')}")
    print(f"{Colors.CYAN}4.{Colors.END} {lang.get('inventory')}")
    print(f"{Colors.CYAN}5.{Colors.END} {lang.get('missions')}")
    print(f"{Colors.CYAN}6.{Colors.END} {lang.get('fight_boss')}")
    print(f"{Colors.CYAN}7.{Colors.END} {lang.get('tavern')}")
    print(f"{Colors.CYAN}8.{Colors.END} {lang.get('shop')}")
    print(f"{Colors.CYAN}9.{Colors.END} {lang.get('alchemy')}")
    print(f"{Colors.CYAN}10.{Colors.END} {lang.get('elite_market')}")
    print(f"{Colors.CYAN}11.{Colors.END} {lang.get('rest')}")
    print(f"{Colors.CYAN}12.{Colors.END} {lang.get('companions')}")
    print(f"{Colors.CYAN}13.{Colors.END} {lang.get('dungeons')}")
    print(f"{Colors.CYAN}14.{Colors.END} {lang.get('challenges')}")

    if player.current_area == "your_land":
        print(
            f"{Colors.CYAN}15.{Colors.END} {lang.get('pet_shop', 'Pet Shop')}")
    print(f"{Colors.CYAN}16.{Colors.END} {lang.get('settings', 'Settings')}")

    if player.current_area == "your_land":
        print(
            f"{Colors.YELLOW}17.{Colors.END} {lang.get('furnish_home', 'Furnish Home')}"
        )
        print(
            f"{Colors.YELLOW}18.{Colors.END} {lang.get('build_structures', 'Build Structures')}"
        )
        print(f"{Colors.YELLOW}19.{Colors.END} {lang.get('farm', 'Farm')}")
        print(
            f"{Colors.YELLOW}20.{Colors.END} {lang.get('training', 'Training')}"
        )
        print(f"{Colors.CYAN}21.{Colors.END} {lang.get('save_game')}")
        print(f"{Colors.CYAN}22.{Colors.END} {lang.get('load_game')}")
        print(f"{Colors.CYAN}23.{Colors.END} {lang.get('claim_rewards')}")
        print(f"{Colors.CYAN}24.{Colors.END} {lang.get('quit')}")
    else:
        print(f"{Colors.CYAN}17.{Colors.END} {lang.get('save_game')}")
        print(f"{Colors.CYAN}18.{Colors.END} {lang.get('load_game')}")
        print(f"{Colors.CYAN}19.{Colors.END} {lang.get('claim_rewards')}")
        print(f"{Colors.CYAN}20.{Colors.END} {lang.get('quit')}")

File: utilities_js/UI.js
/**
 * UI Module for Our Legacy (Browser Version)
 * Terminal-style UI functions
 * Ported from utilities/UI.py
 */

import { Colors } from './settings.js';

/**
 * Clear the terminal screen
 */
export function clearScreen() {
  // Browser version - the game instance should handle clearing
}

/**
 * Create a visual progress bar
 * @param {number} current - Current value
 * @param {number} maximum - Maximum value
 * @param {number} width - Width of the bar
 * @param {string} color - Color code
 * @returns {string} Formatted progress bar
 */
export function createProgressBar(current, maximum, width = 20, color = Colors.GREEN) {
  if (maximum <= 0) {
    return "[" + " ".repeat(width) + "]";
  }
  const filledWidth = Math.floor((current / maximum) * width);
  const filled = "█".repeat(filledWidth);
  const empty = "░".repeat(width - filledWidth);
  const percentage = (current / maximum) * 100;
  return `[${Colors.wrap(filled, color)}${empty}] ${percentage.toFixed(1)}%`;
}

/**
 * Create a visual separator line
 * @param {string} char - Character to use
 * @param {number} length - Length of separator
 * @returns {string} Separator string
 */
export function createSeparator(char = "=", length = 60) {
  return char.repeat(length);
}

/**
 * Create a decorative section header
 * @param {string} title - Section title
 * @param {string} char - Character to use
 * @param {number} width - Width of header
 * @returns {string} Formatted header
 */
export function createSectionHeader(title, char = "=", width = 60) {
  const padding = Math.floor((width - title.length - 2) / 2);
  const headerText = `${char.repeat(padding)} ${title} ${char.repeat(padding)}`;
  return Colors.wrap(headerText, `${Colors.CYAN}${Colors.BOLD}`);
}

/**
 * Display welcome screen
 * @param {Object} lang - Language manager
 * @param {Object} gameInstance - Game instance
 * @param {Function} askFunc - Ask function for input
 * @returns {Promise<string>} User choice
 */
export async function displayWelcomeScreen(lang, gameInstance, askFunc) {
  const game = gameInstance;
  
  while (true) {
    if (game && typeof game.clear === 'function') {
      game.clear();
    }
    game.print(`${Colors.CYAN}${Colors.BOLD}`);
    game.print("=".repeat(60));
    game.print(`             ${lang.get('game_title_display') || 'Our Legacy'}`);
    game.print(`       ${lang.get('game_subtitle_display') || 'A Text-Based Fantasy RPG'}`);
    game.print("=".repeat(60));
    game.print(`${Colors.END}`);
    game.print(lang.get("welcome_message") || "Welcome to Our Legacy!");
    game.print("Choose your path wisely, for every decision shapes your destiny.\n");
    game.print(`${Colors.BOLD}${Colors.CYAN}=== ${lang.get('main_menu') || 'Main Menu'} ===${Colors.END}`);
    game.print(`${Colors.CYAN}1.${Colors.END} ${lang.get('new_game') || 'New Game'}`);
    game.print(`${Colors.CYAN}2.${Colors.END} ${lang.get('load_game') || 'Load Game'}`);
    game.print(`${Colors.CYAN}3.${Colors.END} ${lang.get('settings') || 'Settings'}`);
    game.print(`${Colors.CYAN}4.${Colors.END} ${lang.get('mods') || 'Mods'}`);
    game.print(`${Colors.CYAN}5.${Colors.END} ${lang.get('quit') || 'Quit'}\n`);

    const choice = await askFunc(`${Colors.CYAN}Choose an option (1-5): ${Colors.END}`);
    
    if (choice === "1") {
      return "new_game";
    }
    if (choice === "2") {
      return "load_game";
    }
    if (choice === "3") {
      if (gameInstance && gameInstance.settingsWelcome) {
        await gameInstance.settingsWelcome();
      }
    }
    if (choice === "4") {
      if (gameInstance && gameInstance.modsWelcome) {
        await gameInstance.modsWelcome();
      }
    }
    if (choice === "5") {
      game.print(lang.get("thank_exit") || "Thank you for playing!");
      if (game && typeof game.clear === 'function') {
        game.clear();
      }
      return "quit";
    }
  }
}

/**
 * Display the main game menu
 * @param {Object} lang - Language manager
 * @param {Object} player - Player object
 * @param {string} areaName - Current area name
 * @param {string} menuMax - Maximum menu option number
 * @returns {string} Formatted menu string
 */
export function displayMainMenu(lang, player, areaName, menuMax = "20") {
  let output = "";
  output += `\n${Colors.BOLD}=== ${lang.get('main_menu') || 'Main Menu'} ===${Colors.END}`;
  output += `\n${lang.get("current_location", { area: areaName }) || `Current Location: ${areaName}`}`;

  // Time and weather
  const displayHour = Math.floor(player.hour);
  const displayMinute = Math.floor((player.hour - displayHour) * 60);
  const timeStr = lang.get("current_time", { hour: `${displayHour.toString().padStart(2, '0')}:${displayMinute.toString().padStart(2, '0')}` }) || `Time: ${displayHour.toString().padStart(2, '0')}:${displayMinute.toString().padStart(2, '0')}`;
  const dayStr = lang.get("current_day", { day: player.day.toString() }) || `Day ${player.day}`;
  const weatherDesc = player.getWeatherDescription ? player.getWeatherDescription(lang) : player.currentWeather;
  
  output += `\n${Colors.YELLOW}${timeStr} | ${dayStr}${Colors.END}`;
  output += `\n${Colors.CYAN}${weatherDesc}${Colors.END}`;

  output += `\n${Colors.CYAN}1.${Colors.END} ${lang.get('explore') || 'Explore'}`;
  output += `\n${Colors.CYAN}2.${Colors.END} ${lang.get('view_character') || 'View Character'}`;
  output += `\n${Colors.CYAN}3.${Colors.END} ${lang.get('travel') || 'Travel'}`;
  output += `\n${Colors.CYAN}4.${Colors.END} ${lang.get('inventory') || 'Inventory'}`;
  output += `\n${Colors.CYAN}5.${Colors.END} ${lang.get('missions') || 'Missions'}`;
  output += `\n${Colors.CYAN}6.${Colors.END} ${lang.get('fight_boss') || 'Fight Boss'}`;
  output += `\n${Colors.CYAN}7.${Colors.END} ${lang.get('tavern') || 'Tavern'}`;
  output += `\n${Colors.CYAN}8.${Colors.END} ${lang.get('shop') || 'Shop'}`;
  output += `\n${Colors.CYAN}9.${Colors.END} ${lang.get('alchemy') || 'Alchemy'}`;
  output += `\n${Colors.CYAN}10.${Colors.END} ${lang.get('elite_market') || 'Elite Market'}`;
  output += `\n${Colors.CYAN}11.${Colors.END} ${lang.get('rest') || 'Rest'}`;
  output += `\n${Colors.CYAN}12.${Colors.END} ${lang.get('companions') || 'Companions'}`;
  output += `\n${Colors.CYAN}13.${Colors.END} ${lang.get('dungeons') || 'Dungeons'}`;
  output += `\n${Colors.CYAN}14.${Colors.END} ${lang.get('challenges') || 'Challenges'}`;

  if (player.currentArea === "your_land") {
    output += `\n${Colors.CYAN}15.${Colors.END} ${lang.get('pet_shop', 'Pet Shop')}`;
  }
  output += `\n${Colors.CYAN}16.${Colors.END} ${lang.get('settings', 'Settings')}`;

  if (player.currentArea === "your_land") {
    output += `\n${Colors.YELLOW}17.${Colors.END} ${lang.get('furnish_home', 'Furnish Home')}`;
    output += `\n${Colors.YELLOW}18.${Colors.END} ${lang.get('build_structures', 'Build Structures')}`;
    output += `\n${Colors.YELLOW}19.${Colors.END} ${lang.get('farm', 'Farm')}`;
    output += `\n${Colors.YELLOW}20.${Colors.END} ${lang.get('training', 'Training')}`;
    output += `\n${Colors.CYAN}21.${Colors.END} ${lang.get('save_game') || 'Save Game'}`;
    output += `\n${Colors.CYAN}22.${Colors.END} ${lang.get('load_game') || 'Load Game'}`;
    output += `\n${Colors.CYAN}23.${Colors.END} ${lang.get('claim_rewards') || 'Claim Rewards'}`;
    output += `\n${Colors.CYAN}24.${Colors.END} ${lang.get('quit') || 'Quit'}`;
  } else {
    output += `\n${Colors.CYAN}17.${Colors.END} ${lang.get('save_game') || 'Save Game'}`;
    output += `\n${Colors.CYAN}18.${Colors.END} ${lang.get('load_game') || 'Load Game'}`;
    output += `\n${Colors.CYAN}19.${Colors.END} ${lang.get('claim_rewards') || 'Claim Rewards'}`;
    output += `\n${Colors.CYAN}20.${Colors.END} ${lang.get('quit') || 'Quit'}`;
  }

  return output;
}

export default {
  clearScreen,
  createProgressBar,
  createSeparator,
  createSectionHeader,
  displayWelcomeScreen,
  displayMainMenu
};
